llvm/lib/Transforms/Vectorize/VPlan.cpp

```diff
@@ -934,8 +934,19 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
 
   IRBuilder<> Builder(State.CFG.PrevBB->getTerminator());
   // FIXME: Model VF * UF computation completely in VPlan.
-  VFxUF.setUnderlyingValue(
-      createStepForVF(Builder, TripCountV->getType(), State.VF, State.UF));
+  if (VF.getNumUsers()) {
+    Value *RuntimeVF =
+        createStepForVF(Builder, TripCountV->getType(), State.VF, 1);
+    VF.setUnderlyingValue(RuntimeVF);
+    VFxUF.setUnderlyingValue(
+        State.UF > 1
+            ? Builder.CreateMul(
+                  RuntimeVF, ConstantInt::get(TripCountV->getType(), State.UF))
+            : RuntimeVF);
+  } else {
+    VFxUF.setUnderlyingValue(
```

Comment:
I see earlier that
> ...  Code for VF is only generated if there are users of VF, to avoid unnecessary test changes.
Code for VF must be generated regardless of direct users, issues is its position and possible repetition?

Reply:
Yes, `VFxUF` is used to increment the canonical induction, which is present in almost all cases, but could also be done only if users exist.

> ... Code for VF is only generated if there are users of VF, to avoid unnecessary test changes.
> Code for VF must be generated regardless of direct users, issues is its position and possible repetition?

There are cases where VF separately isn't used, only as part of `VFxUF`. If only VFxUF is used, the constant multiply of VF * UF is folded, hence always generating VF separately would lead to additional test changes. Some alternatives are mentioned in my latest comment above.

Reply:
At the moment, it is always used (added assert in 1a5a1e97817).